File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
	long long N, M;
	cin >> N >> M;

	vector<long long> trees(N);
	for (int i = 0; i < N; i++) {
		cin >> trees[i];
	}

	sort(trees.begin(), trees.end());

	long long start = 0;
	long long end = trees[N - 1]; // 가장 높은 나무

	long long ans = 0;

	while (start <= end) {
		long long mid = (start + end) / 2;

		long long sum = 0;
		for (int i = 0; i < N; i++) {
			if (trees[i] > mid)
				sum += (trees[i] - mid);
		}

		if (sum >= M) {
			ans = mid;
			start = mid + 1;
		}
		else {
			end = mid - 1;
		}
	}

	cout << ans;

	return 0;
}

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] trees = new int[N];

        long treeLen = 0;
        int left = 0;
        int right = 0;
        int ans = 0;

        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++){
            trees[i] = Integer.parseInt(st.nextToken());
            right = Integer.max(right, trees[i]);
        }


        while (left < right){
            int mid = (left + right) / 2;
            treeLen = 0;
            for(int i = 0; i < N; i++){
                treeLen += Integer.max(0, trees[i] - mid);
            }
            if(treeLen >= M){
                ans = mid;
                left = mid + 1;
            }
            else{
                right = mid;
            }
        }

        System.out.println(ans);
    }
}